### PR TITLE
[serve] Allow pluggable router classes via handle config (experimental)

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Add a way to configure the router via the handle config. This can be used to implement custom routing strategies. It may be better in the long term to make the routing configured via the deployment decorator.

TODO:
- [x] Add unit test

### CODE REVIEW

File: python/ray/serve/_private/router.py [hunk 1]
```diff
@@ -894,14 +895,18 @@ def __init__(
         deployment_name: str,
         event_loop: asyncio.BaseEventLoop = None,
         _use_new_routing: bool = False,
+        _router_cls: str = "",
```
Comment (edoakes) [inline python/ray/serve/_private/router.py:898]: nit: 
```suggestion
        _router_cls: Optional[str] = None,
```

File: python/ray/serve/_private/router.py [hunk 2]
```diff
@@ -894,14 +895,18 @@ def __init__(
...
         """Used to assign requests to downstream replicas for a deployment.
 
         The scheduling behavior is delegated to a ReplicaScheduler; this is a thin
         wrapper that adds metrics and logging.
         """
         self._event_loop = event_loop
-        if _use_new_routing:
+
+        if _router_cls:
+            self._replica_scheduler = _load_class(_router_cls)()
```
Comment (edoakes) [inline python/ray/serve/_private/router.py:908]: can you add a log statement indicating the class name used similar to below. or combine them all into one statement like:
```python
logger.info(f"Using {self._replica_scheduler.__class__}.", extra={"log_to_stderr": False})
```

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -28,6 +28,7 @@
 from ray.dag.py_obj_scanner import _PyObjScanner
 from ray.exceptions import RayActorError, RayTaskError
 from ray.util import metrics
+from ray._private.storage import _load_class
```
Comment (sihanwang41) [inline python/ray/serve/_private/router.py:31]: is this `storage` from data team lib?

**Review by edoakes — COMMENTED** [review]
looks good pending test

File: python/ray/serve/_private/router.py (hunk 3 above)
Comment (ericl) [reply python/ray/serve/_private/router.py:31]: It's common, but I moved it to utils.py for clarity.

File: python/ray/serve/handle.py [hunk 4]
```diff
@@ -175,20 +181,22 @@ def _options(
...
     ):
         new_handle_options = self.handle_options.copy_and_update(
             method_name=method_name,
             multiplexed_model_id=multiplexed_model_id,
             stream=stream,
+            _router_cls=_router_cls,
         )
 
-        if self._router is None:
+        if self._router is None and not _router_cls:
```
Comment (ericl) [inline python/ray/serve/handle.py:193]: Alternatively, I could move this to construct the router in the handle constructor. However, I didn't know if this deferred construction was intentional.

**Review by ericl — COMMENTED** [review]
ptal, test added

**Review by sihanwang41 — APPROVED** [review]

**Pull request merged**